src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs

```diff
@@ -217,19 +290,80 @@ protected string ContainingDirectoryPathOpt
             }
         }
 
+        #region ProjectFilePath
+        private string _projectFilePath;
+        
         /// <summary>
         /// The full path of the project file. Null if none exists (consider Venus.)
+        /// Note that the project file path might change with project file rename.
+        /// If you need the folder of the project, just use <see cref="ContainingDirectoryPathOpt" /> which doesn't change for a project.
         /// </summary>
-        public string ProjectFilePath => _filePathOpt;
+        public string ProjectFilePath
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    return _projectFilePath;
+                }
+            }
+            private set
+            {
+                lock (_gate)
+                {
+                    _projectFilePath = value;
+                }
+            }
+        }
+        #endregion
+
+        #region DisplayName
+        private string _displayName;
 
         /// <summary>
         /// The public display name of the project. This name is not unique and may be shared
         /// between multiple projects, especially in cases like Venus where the intellisense
         /// projects will match the name of their logical parent project.
         /// </summary>
-        public string DisplayName { get; protected set; }
+        public string DisplayName
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    return _displayName;
+                }
+            }
+            private set
+            {
+                lock (_gate)
+                {
+                    _displayName = value;
+                }
+            }
+        }
+        #endregion
 
-        internal string AssemblyName { get; private set; }
+        #region AssemblyName
+        private string _assemblyName;
+        internal string AssemblyName
+        {
+            get
+            {
+                lock (_gate)
+                {
+                    return _assemblyName;
+                }
+            }
+            private set
+            {
+                lock (_gate)
+                {
+                    _assemblyName = value;
+                }
```

Comment:
I question the value of the majority of locks you added in property implementations.  Can you detail why they are needed?  

The values here are all reference types hence we already know the assignments are atomic.  The presence of `lock` is only useful when you need to synchronize those read / writes with other operations.  I don't see that happening here.